### PR TITLE
Add Electron example

### PR DESCRIPTION
**Pull request by sergiodxa**
This example show how you can use Next.js inside an Electron application to avoid a lot of configuration, have routes and use server-render to speed up the initial render of the application.

### CODE REVIEW

**Review by matheuss — COMMENTED** [review]
I'm not sure if it's a good idea to fire up a HTTP server inside an app like that – it can introduce huge security vulnerabilities.

I'd wait for `next export` 🤔

**Comment by sergiodxa** [comment]
@matheuss I think it depends on how `next export` it's going to work, if it's going to only generate static html files without client-side `getInitialProps` then it's not going to be useful for this use case 🤔 also you will need to re-export your app after every change

BTW, what kind of security vulnerabilities do you thinks can this introduce?

File: examples/with-electron/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+```
+
+Install it and run:
+
+```bash
+npm install
+npm run dev
+```
+
+Deploy it to the cloud with [now](https://zeit.co/now) ([download](https://zeit.co/download))
```
Comment (frol) [inline examples/with-electron/README.md]: Are these instructions relevant to this example?

Comment (sergiodxa) [inline examples/with-electron/README.md]: good point 😅

**Comment by 1msoft** [comment]
good idea, share some pages with some friends.

**Comment by rauchg** [comment]
@sergiodxa I think @matheuss is concerned potentially about browsers probing `localhost`, finding the port, then finding open routes? Maybe an advisory in the README about that vector would be good

**Comment by rauchg** [comment]
`next export` will certainly output 100% static stuff. So, if you only limit yourself to serving static stuff via Next, there shouldn't be any security issues. Protecting `POST` stuff is the key, and not performing side effects upon `GET`

**Comment by sergiodxa** [comment]
In browsers the client code breaks (because Webpack doesn't do anything to the `require` function), you only get the server rendered page. Anyway I added an warning in the README about the possible access from a browser.

I also added a little validation code to respond with 404 if the request does not come from an Electron app. It's not fully secure but it should hide, for unintentional users, the access to the app from a browser.

File: examples/with-electron/package.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "scripts": {
+    "dev": "electron index.js",
+    "build": "next build",
+    "start": "NODE_ENV=production electron index.js"
+  },
+  "dependencies": {
+    "electron": "1.6.2",
+    "next": "beta",
```
Comment (frol) [inline examples/with-electron/package.json]: I believe it is worth pinning Next.js version as it is done in other examples:

```
"next": "^2.0.0-beta",
```

File: examples/with-electron/package.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    "react": "15.4.2",
+    "react-dom": "15.4.2"
```
Comment (frol) [inline examples/with-electron/package.json]: Other examples pin React version like this:

```
"react": "^15.4.2",
"react-dom": "^15.4.2"
```

**Comment by rauchg** [comment]
@sergiodxa can you look into whether a unix socket would be a viable option?

https://nodejs.org/api/net.html#net_server_listen_path_backlog_callback

**Comment by leo** [comment]
[Now Desktop](https://github.com/zeit/now-desktop) is another example now.

**Comment by sergiodxa** [comment]
@rauchg I have tried to use a Unix Socket with the net module but Next require an request and response object, I'm not use if it's possible : thinking:

@leo yeah! I just checked the code and the server it's basically the same, awesome, using Next inside Electron to avoid configuration it's pure magic.

I also published yesterday a [Spanish tutorial about this example](https://platzi.com/blog/crea-aplicaciones-de-escritorio-con-electron-y-nextjs/).

**Comment by leo** [comment]
Please wait with this. I'll release a blog post about this very soon (describes how to use Next.js with Electron in detail)... 😊

**Comment by ritz078** [comment]
@sergiodxa Have you tested by creating a package ? I created a darwin build and it didn't launch.

**Comment by leo** [comment]
We're using it [here](https://github.com/zeit/now-desktop). Feel free to test!

**Comment by sergiodxa** [comment]
@ritz078 Yes, I tested it using this example, are you having any error in console?

**Comment by ritz078** [comment]
No there isn't any error in the console.

I cloned the same and added `electron-packager`

```js
{
  "scripts": {
    "dev": "electron index.js",
    "build": "next build",
    "start": "NODE_ENV=production electron index.js",
    "pack": "electron-packager . --out=package --darwin --overwrite"
  },
  "dependencies": {
    "electron": "1.6.2",
    "next": "^2.0.0-beta",
    "react": "^15.4.2",
    "react-dom": "^15.4.2"
  },
  "devDependencies": {
    "electron-packager": "^8.6.0"
  }
}
```
The app is being created fine `npm run pack` but its not launching the window.

**Comment by ritz078** [comment]
It seems this is the error.

```js

// written here executes
nextApp
  .prepare()
  .then(() => {
// anything written here doesn't execute after packaging
})
```

**Comment by sergiodxa** [comment]
@ritz078 I'm checking this, I'm going to bring some things from now-desktop that can be useful to easily pack the application

**Comment by ritz078** [comment]
@sergiodxa  That will be great. Thanks.

**Comment by trevordmiller** [comment]
Would be neat to see how to use Next.js in Electron; @leo is that post available somewhere I can read?

**Comment by leo** [comment]
@trevordmiller I'm working on it right now. Should be up very soon! 😊

**Comment by trevordmiller** [comment]
I've ported a working client-side app to electron app using next, following the now-desktop example from @leo:
https://github.com/trevordmiller/hideaway

It's close to working but I'm running into two new issues once electron + next were introduced as shown in the gif below:
1) the client doesn't render on first load, but requires a refresh
2) class instances are not being bound
next-electron-issues

As this may be applicable to this PR or any electron/next.js documentation, does anyone have any ideas what could be wrong here? Thank you.

**Comment by trevordmiller** [comment]
^ The `body` is empty on initial render:
image

**Comment by ritz078** [comment]
@trevordmiller I am using `babel-plugin-transform-decorators-legacy` and using named methods. Maybe this can solve your problem.  https://github.com/ritz078/snape

**Comment by trevordmiller** [comment]
@ritz078 Sorry but I don't understand what decorators have to do with the initial render not populating the dom? And I'm using named methods: https://github.com/trevordmiller/hideaway

**Comment by trevordmiller** [comment]
Per my issue, it was related to Next and would be helpful to add to the example documentation. @sergiodxa helped me in the Zeit Slack:

> "when you run Next in a custom server you do `nextApp.prepare().then(configureRoutes)` right? in that `then` you should be opening the Electron window. it's not perfect but [check this](https://github.com/sergiodxa/next.js/blob/2d7ec1cca9a752696b0b74c181f385a23a403b14/examples/with-electron/index.js). I had the same problem when trying to program that example. if you open the Electron window *before* finishing the Next.js prepare (build files en dev) then it open an empty window. but if you reload or open a new window it renders well. in development it's going to be slow (because of build time), but in production is fast"

**Comment by sergiodxa** [comment]
@trevordmiller in this PR example it's doing that, maybe it's a good idea to comment why

**Comment by trevordmiller** [comment]
@sergiodxa @leo Here is my commit diff to get the initial render working if it is helpful for this PR or the blog post: https://github.com/trevordmiller/hideaway/commit/4b20a8e6af430e640cfbe5db996c69da20d27717

**Comment by sergiodxa** [comment]
I updated this example to use Next.js v3 and use `next export` to generate HTML files.

**Comment by sergiodxa** [comment]
https://github.com/PlatziDev/pulse
👆 an electron app using Next v3 :)

**Comment by frol** [comment]
@sergiodxa `name`, `description`, `version`, `author` fields of package.json are required for `npm run dist`.

Thank you for the example! It works great!

**Comment by leo** [comment]
I discovered that `next export` doesn't yet work with Now Desktop. But we're working on making it possible and then the post is the next step!

**Comment by leo** [comment]
Now Desktop is now using `next export`. This means my post is coming very soon! 🤞

**Comment by frol** [comment]
@sergiodxa I have given this example a try on a bigger project and bumped into an issue with `superagent`. Once I import `superagent` I get this error in the Electron Console in dev mode (production build renders the initial HTML, but JS seems to fail):

```
incoming_form.js?4115e37:3 Uncaught TypeError: require is not a function
```

Any ideas?

**Comment by sergiodxa** [comment]
@frol I think that's a library specific problem, not an Electron o Next.js problem, I have no problems with other libraries 🤔

**Comment by frol** [comment]
@sergiodxa Yes, it was a `formidable` specific issue.

I have tested the example further and it turns out that links between pages work correctly only on the root page in the exported mode (`/ -> /1` works fine, but `/1 -> /` doesn't). It is due to the use of `assetPrefix: './'`. It tries to load the JS files relatively to the current URL path... I am digging into it further.

**Comment by frol** [comment]
@sergiodxa I have figured it out with a help of Now-Desktop example. They use `electron-routes`. I will prepare a PR against your fork in a few minutes once I test it.

**Comment by sergiodxa** [comment]
@frol yeah they use `electron-routes` because they use next v2 without `next export`, I have used the `assetPrefix: './'` because without it when you export it doesn't load the JS files.

**Comment by frol** [comment]
@sergiodxa Well, `electron-routes` didn't play well in the end. You are right, we should use the exported files. I will give it another try.

**Comment by frol** [comment]
@sergiodxa I have figured it out with a help of a custom Electron protocol: https://electron.atom.io/docs/api/protocol/#protocol

I am testing it right now and will create a PR.

**Comment by frol** [comment]
@sergiodxa For some really weird reason, I cannot create a PR against your fork. Please, pull my commit: https://github.com/frol/next.js/commit/fd749a228d9f88d09a40e3cce7cf9ddac995f178

**Comment by frol** [comment]
By the way, you cannot use cookies for "current domain" when your page is open as `file://`. I have experimented with using the custom `next://` protocol, but it doesn't play nicely as well. I may simply use other means for the storage as cookies don't make a lot of sense in a standalone app, but it just has different behavior from the dev mode, so it is better to be prepared.

P.S. I have a biggish app in Next.js, and I am just stressing this Electron example with it.

**Comment by frol** [comment]
I have found a workaround for the cookies in Electron, just use [@expo/electron-cookies](https://github.com/expo/electron-cookies). Here is a snippet which you need to put somewhere in the top-component (`_document.js` or common high-level component):

```javascript
if (typeof document !== 'undefined') {
  const ElectronCookies = require('@exponent/electron-cookies')
  ElectronCookies.enable({
    origin: 'https://example.com',
  });
}
```

So far, I have managed to package my whole Next.js application into Electron and it is fully functioning! I had to only workaround the superagent/node-formidable issue, the missing cookies support and files upload implementation (I had to use `fs.createReadStream()` instead of plain File objects when POSTing them).

@sergiodxa Thank you for the example!

**Comment by trevordmiller** [comment]
@leo Any updates on the blog post? I'm looking forward to it :) Thank you for the examples so far.

**Comment by leo** [comment]
@trevordmiller Coming very soon - sadly can't tell anything else publicly. But it's VERY close!

File: examples/with-electron/renderer/next.config.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    return config
+  },
+  exportPathMap () {
+    // export our pages as HTML for production usage
+    return {
+      '/': { page: '/' }
+    }
+  },
+  // set the prefix as `./` instead of `/`, this is because when you export your pages
```
Comment (frol) [inline examples/with-electron/renderer/next.config.js]: The comment does not reflect the actual setting. There is no `./` anymore.

Comment (sergiodxa) [inline examples/with-electron/renderer/next.config.js]: Thanks I fixed it :)

File: examples/with-electron/main/set-menu.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    ]
+  })
+
+  // Edit menu
+  template[1].submenu.push(
+    { type: 'separator' },
+    {
+      label: 'Speech',
+      submenu: [{ role: 'startspeaking' }, { role: 'stopspeaking' }]
+    },
```
Comment (frol) [inline examples/with-electron/main/set-menu.js]: I had to remove the comma since otherwise it fails with "SyntaxError: Unexpected token )".

Comment (sergiodxa) [inline examples/with-electron/main/set-menu.js]: Fixed it, thanks a lot for your feedback :)

**Comment by frol** [comment]
@sergiodxa Thank you for the updated example! It is really easy to get any Next.js project packaged into Electron now!

**Comment by timneutkens** [comment]
@leo what's the status of this one? 😄

**Comment by ragingwind** [comment]
I'm also waiting for this to introduce to our devs as an organizer of Electron meetup in Korea and a big fan of next.js

**Comment by leo** [comment]
My [helper package](https://github.com/leo/electron-next) is already out! Blog post very close... 🚀

**Comment by timneutkens** [comment]
@leo cool 👍 🤗 🚀 :shipit:

**Comment by leo** [comment]
Just published [my blog post](https://twitter.com/notquiteleo/status/886733954006601729) about building Electron apps using Next.js and created a [boilerplate](https://github.com/leo/electron-next-skeleton). If it looks good to you, I suggest basing this PR on it! 🚀 ❤️

**Comment by trevordmiller** [comment]
Thanks @leo 🎉

**Comment by sergiodxa** [comment]
I updated this example to bring some ideas of electron-next-skeleton, now it's simpler, but I keep the UI so the example show how you can import Electron modules in the renderer process and use IPC between the Next.js (renderer) app and the main process.

**Review by timneutkens — APPROVED** [review]
Looks good ❤️
